cmd/dockerd: add basic support for sd_notify "reload" notifications

thaJeztah opened the pull request:
- relates to https://github.com/moby/moby/pull/22446
- addresses https://github.com/moby/moby/issues/47353


commit f74b856e1ac2805fe48ceb52bc83cd7a3cec870c added an ExecReload to the systemd unit, which allows users to signal the daemon to reload its config through systemd (`systemctl reload docker.service`).

While reloading works, systemd expects the `ExecReload` command to be synchronous, so that it knows when the reload completes, and can account for this when managing dependent services.

> Note however that reloading a daemon by enqueuing a signal (...) is usually
> not a good choice, because this is an asynchronous operation and hence not
> suitable when ordering reloads of multiple services against each other.

Systemd 253 introduced a new Type (Type=notify-reload, see [systemd#25916]), which allows setting a "ReloadSignal" instead, in addition to sending a [MONOTONIC_USEC].

We currently still support distros that do not provide systemd 253, so cannot use this new feature, but sending "RELOADING=1 / "READY=1" should at least provide more information to systemd for the time being.

This patch:

- adds reload notifications to the daemon to notify when the daemon got signaled to reload its configuration see [sd_notify(3)].
- adds a [notifyReady] to notify when the reload finished, which we currently send regardless if the reload was successful or failed (which could be due to an invalid config).

We can re-implement this using `Type=notify-reload` once we no longer have to support systemd versions before 253.

[sd_notify(3)]: https://www.freedesktop.org/software/systemd/man/latest/sd_notify.html#RELOADING=1
[systemd.service(5)]: https://www.freedesktop.org/software/systemd/man/latest/systemd.service.html#Type=
[ExecReload]: https://www.freedesktop.org/software/systemd/man/latest/systemd.service.html#ExecReload=
[MONOTONIC_USEC]: https://www.freedesktop.org/software/systemd/man/latest/sd_notify.html#MONOTONIC_USEC=…
[systemd#25916]: https://github.com/systemd/systemd/pull/25916

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

**- How I did it**

**- How to verify it**

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->
```
- Send sd_notify ["READY"](https://www.freedesktop.org/software/systemd/man/latest/sd_notify.html#READY=1) and ["STOPPING"](https://www.freedesktop.org/software/systemd/man/latest/sd_notify.html#STOPPING=1) synchronously to make sure they are sent before we proceed.
- Add sd_notify ["RELOADING"](https://www.freedesktop.org/software/systemd/man/latest/sd_notify.html#RELOADING=1) notifications when signalling the daemon to reload its configuration.
```



**- A picture of a cute animal (not mandatory but encouraged)**

thaJeztah committed:
cmd/dockerd: add basic support for sd_notify "reload" notifications

commit f74b856e1ac2805fe48ceb52bc83cd7a3cec870c added an ExecReload to the
systemd unit, which allows users to signal the daemon to reload its config
through systemd (`systemctl reload docker.service`).

While reloading works, systemd expects the `ExecReload` command to be
synchronous, so that it knows when the reload completes, and can account
for this when managing dependent services.

> Note however that reloading a daemon by enqueuing a signal (...) is usually
> not a good choice, because this is an asynchronous operation and hence not
> suitable when ordering reloads of multiple services against each other.

Systemd 253 introduced a new Type (Type=notify-reload, see [systemd#25916]),
which allows setting a "ReloadSignal" instead, in addition to sending a
[MONOTONIC_USEC].

We currently still support distros that do not provide systemd 253, so cannot
use this new feature, but sending "RELOADING=1 / "READY=1" should at least
provide more information to systemd for the time being.

This patch:

- adds reload notifications to the daemon to notify when the daemon got
  signaled to reload its configuration see [sd_notify(3)].
- adds a [notifyReady] to notify when the reload finished, which we currently
  send regardless if the reload was successful or failed (which could be
  due to an invalid config).

We can re-implement this using `Type=notify-reload` once we no longer have to
support systemd versions before 253.

[sd_notify(3)]: https://www.freedesktop.org/software/systemd/man/latest/sd_notify.html#RELOADING=1
[systemd.service(5)]: https://www.freedesktop.org/software/systemd/man/latest/systemd.service.html#Type=
[ExecReload]: https://www.freedesktop.org/software/systemd/man/latest/systemd.service.html#ExecReload=
[MONOTONIC_USEC]: https://www.freedesktop.org/software/systemd/man/latest/sd_notify.html#MONOTONIC_USEC=…
[systemd#25916]: https://github.com/systemd/systemd/pull/25916

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

corhere commented on cmd/dockerd/daemon_linux.go:
```diff
@@ -35,6 +35,37 @@ func notifyReady() {
+// [systemd#25916]: https://github.com/systemd/systemd/pull/25916
+func notifyReloading() {
+	// TODO(thaJeztah): use "Type=notify-reload", "ReloadSignal", and "MONOTONIC_USEC" once we drop support for systemd < 253, and supported by github.com/coreos/go-systemd.
+	go systemdDaemon.SdNotify(false, systemdDaemon.SdNotifyReloading)
```
```suggestion
	systemdDaemon.SdNotify(false, systemdDaemon.SdNotifyReloading)
```

Ditto for `notifyReady()`. Notifying asynchronously makes it possible for the two notifications to race, and possibly get delivered to systemd out of order.

thaJeztah replied:
> Notifying asynchronously makes it possible for the two notifications to race, and possibly get delivered to systemd out of order.

I was considering that as well yeah. Do you think we should call them synchronously?

corhere reviewed (commented):
Sounds good to me, in principle

neersighted commented on cmd/dockerd/daemon_linux.go:
```diff
@@ -35,6 +35,37 @@ func notifyReady() {
+// [MONOTONIC_USEC]: https://www.freedesktop.org/software/systemd/man/latest/sd_notify.html#MONOTONIC_USEC=…
+// [systemd#25916]: https://github.com/systemd/systemd/pull/25916
+func notifyReloading() {
+	// TODO(thaJeztah): use "Type=notify-reload", "ReloadSignal", and "MONOTONIC_USEC" once we drop support for systemd < 253, and supported by github.com/coreos/go-systemd.
```
The parts of this comment relating to the unit (`Type` and `ReloadSignal`) should go in that unit and not here.

thaJeztah replied:
I tried to avoid cluttering the systemd unit with TODO's, so thought mentioning here that both need to be updated once we change would give a bit of context of the scope of changes.

I removed mention of the `Type` and `ReloadSignal`, and added a link to this PR/comment instead.

thaJeztah committed:
cmd/dockerd: don't make notifyReady(), notifyStopping() asynchronous

Leave it to the consumer to decide whether these should be called in a
goroutine.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented on cmd/dockerd/daemon.go:
```diff
@@ -477,6 +477,16 @@ func (cli *DaemonCli) reloadConfig() {
+	//
+	// We could make this more granular (i.e., don't notify if initial
+	// validation failed, and we aborted before actually reloading config.
+	go notifyReloading()
+	defer func() { go notifyReady() }()
```
I made handling these (and the existing ones) asynchronously a choice for the caller, but perhaps we should consider indeed to not making these asynchronous; https://github.com/moby/moby/pull/47358#discussion_r1486179414

@neersighted @corhere WDYT?

neersighted replied:
The intention seems to be avoiding blocking the caller with an "optional" task like this, but I wonder if this is too lightweight given there are now synchronization issues at play, and perhaps we should instead use a channel + background goroutine to coordinate these calls?

thaJeztah replied:
Yeah, honestly not sure how heavy-weight these are. Perhaps it's fine to just remove the `go` here, and just call them synchronously 🤔

corhere replied:
I think it'd be fine to call them synchronously. [They open a UNIX datagram socket, write and close.](https://github.com/coreos/go-systemd/blob/d5623bf85e8e73ae6352f78ee6b55a287619dd4e/daemon/sdnotify.go#L48-L84) I highly doubt the calls would block for a significant length of time unless pid 1 has become unresponsive, in which case you'd have much bigger problems. It'd have to be lightweight and fast as I can't imagine many C programs would be going through the trouble of invoking `sd_notify()` in a background thread. (Note that the syslog API is very similar: apps log by sending unixgrams to the `/dev/log` socket.)

corhere commented on cmd/dockerd/daemon_linux.go:
```diff
@@ -32,12 +32,43 @@ func preNotifyReady() {
+// [MONOTONIC_USEC]: https://www.freedesktop.org/software/systemd/man/latest/sd_notify.html#MONOTONIC_USEC=…
+// [systemd#25916]: https://github.com/systemd/systemd/pull/25916
+func notifyReloading() {
+	// TODO(thaJeztah): Set "MONOTONIC_USEC" once we drop support for systemd < 253, and supported by github.com/coreos/go-systemd, and update systemd unit accordingly; see https://github.com/moby/moby/pull/47358#discussion_r1483533255.
+	_, _ = systemdDaemon.SdNotify(false, systemdDaemon.SdNotifyReloading)
```
- https://github.com/coreos/go-systemd/pull/435

That was a little bit annoying to write; it had to be compatible with _Go 1.12!_ I wonder how long that PR will be left to rot before it gets reviewed...

corhere commented:
> We currently still support distros that do not provide systemd 253, so cannot use [Type=notify-reload]

[As I commented above,](https://github.com/moby/moby/pull/47358#discussion_r1488129879) the daemon could implement everything necessary to support `Type=notify-reload` without breaking compatibility with older versions of systemd as the sd_notify API is forwards-compatible. The only compatibility concern is with the unit file itself. But that's just a packaging concern. Package authors for systemd 253+ distros could ship a modified unit file, or users could create a drop-in for `Type=notify-reload`. We could even provide multiple systemd unit files compatible with different systemd versions for distro-package authors to pick from!

I see no reason to wait until we drop support for systemd < 253 to make the daemon fully compatible with notify-reload, irrespective of whether the unit files shipped in Docker CE packages leverage such compatibility.

corhere commented on cmd/dockerd/daemon.go:
```diff
@@ -477,6 +477,17 @@ func (cli *DaemonCli) reloadConfig() {
+	//
+	// We could make this more granular (i.e., don't notify if initial
+	// validation failed, and we aborted before actually reloading config.
+	if notifyReloading() {
+		defer func() { notifyReady() }()
+	}
```
Using a data dependency on the `notifyReloading` return value to express through the API that the `notifyReady` call needs to be synchronized with it is a brilliant design choice. I suggest going one step further: have `notifyReloading` return a closure which notifies that the reload is done when called. That would have the dual benefit [of making the API both easier to use correctly and harder to misuse](https://blog.codinghorror.com/falling-into-the-pit-of-success/), and of future-proofing the API against other reload-notification protocols we may want to support.
```go
func notifyReloading() (done func())
```
```suggestion
	defer notifyReloading()()
```

thaJeztah replied:
Ah, I had to run after pushing, and wanted to comment that as alternative, but didn't have the time. Looks like we had the same idea here; I'll update to use that approach 🥹👍

thaJeztah commented on cmd/dockerd/daemon.go:
```diff
@@ -477,6 +477,17 @@ func (cli *DaemonCli) reloadConfig() {
+	// but must signal "ready" after this completes (even on failure), which
+	// is done by the reload function defined above.
+	//
+	// We could make this more granular (i.e., don't notify if initial
+	// validation failed, and we aborted before actually reloading config.
```
Ah, that's a good point. I didn't consider that systemd may be checking if its signal was received in good order here.


My train of thought here was to keep the "reloading" state as short as possible. Thinking that systemd would only need to take the _actual_ re-configuring into account; _"hold your horses: we're re-configuring"_, and if no re-configuration happened (validation failed), nothing changed in that perspective, so no need to tell systemd.

I'll remove this snippet from the comment.

thaJeztah committed:
cmd/dockerd: send notifyReady, notifyStopping synchronous

These were called in a goroutine, likely to prevent them from affecting
performance, but doing so also means that there's no promise that the
notifications were actually sent before we continue.

Summarising the discussion from [PR 47358]:

These calls should not be particularly heavy-weight; They open a UNIX datagram
socket, write and close. (see the [go-systemd] implementation), and it's
unlikely these calls would block for a significant length of time unless pid 1
has become unresponsive, in which case you'd have much bigger problems. These
should be lightweight and fast as we can't imagine many C programs would be going
through the trouble of invoking `sd_notify()` in a background thread. Note that
the syslog API is also very similar: apps log by sending unixgrams to the `/dev/log`
socket.

[go-systemd]: https://github.com/coreos/go-systemd/blob/d5623bf85e8e73ae6352f78ee6b55a287619dd4e/daemon/sdnotify.go#L48-L84
[PR 47358]: https://github.com/moby/moby/pull/47358#discussion_r1488177590

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah reviewed (commented):
Updated to make `notifyReloading` return a func.
